### DISCUSSION TITLE
Modified version of GPLv3 licensed source code with additional MIT protected source code.

### DISCUSSION THREAD

**Discussion by zeleny-ondrej** [community/community#29042] [Other Feature Feedback, Questions, & Ideas] [answered]
Hi,

I've use a GPLv3 licensed project as a base for my work, i somewhat modified it and added two more source code files which are under MIT license (also modified).

How do i properly labels this project so i dont brake any license rules.

so far, i've added two licence text files(one MIT license text with authors name, and GNU GPL-V3 license text) and labeled all the files from the original GPLv3 project as GPLv3 licensed. From what i read, i have to put the whole project as GPLv3 licensed, but what about the MIT licensed files ? What else do i have to disclaim ?

**Comment by airtower-luna** [comment, marked as answer]
Disclaimer: I'm not a lawyer, so don't read the following as legal advice. I've just been working on and with FOSS for a while and am writing about how I've seen things done.

> i have to put the whole project as GPLv3 licensed, but what about the MIT licensed files ?

You have to put the *combined work* under GPL. What exactly that means depends a bit on your project, for example you could have multiple source files with a mix of GPL and MIT licenses that need to be compiled into a binary. In that case the *combined work* is the binary, while each source file can keep its own license. Mind that this only works because the MIT license (unlike the GPL, whether v2 or v3) allows distribution of binaries under a different license.

That is how the Linux kernel does it, for example certain drivers are under permissive licenses (similar to the MIT license). Here's one example: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/net/wireless/ath/ath.h?id=144cd24dbc36650a51f7fe3bf1424a1432f1f480

It's critical to keep copyright and license notes in each file intact, both licenses require that. For the GPL a license file in the repository is important. The MIT license is usually written out in full in the licensed files anyway (but a license file is still helpful). Giving some hints in the documentation as to which license applies to which part of your project (so users don't get confused and have to dig through the code to find out) is polite.

**Reply by zeleny-ondrej** [reply]
Thank you for your answer.

I should furtner note that i completly reorganized the function and classes into new files so they do not contain any disclamers beside the mention of "GPL-V3 license" or "MIT license" that i've added.

So from what i understood from your answer is that what i need to do is:

1.  add GLP licence file ( I've already copied it from the original project)
2.  at top of each file mention which license applies to them and probably note that the project is derivated from ..... +link and maybe autor ?

and the project should be fine ?

**Reply by airtower-luna** [reply]
I'm afraid you're in a bit of trouble then. Both [MIT license](https://choosealicense.com/licenses/mit/) and [GPLv3](https://choosealicense.com/licenses/gpl-3.0/) strictly require *preserving* the copyright and license notes. If you use the code without that (let alone other requirements of the GPL) that's a copyright violation. I strongly encourage you to read the actual licenses. The MIT license is really short, for the GPLv3 section 4 and 5 describe that requirement.

I have reorganized code into new files *under the same license* before, here's one example: [`gnutls_io.h`](https://github.com/airtower-luna/mod_gnutls/blob/3855d9ee2461fba7fa75429260e0d714a1cfd694/src/gnutls_io.h) Note that this is "how I've done it and no-one complained", it is **not** "lawyer approved and court-tested". In practice most FOSS developers are happy to assume good faith about such things and aren't going to sue about oddly phrased notices about reorganization etc., but I'd stop assuming good faith if copyright and license notes have been removed.

Mixing code under different licenses in one file is a bigger mess than I'd be willing to trouble myself with. :sweat_smile:

**Reply by zeleny-ondrej** [reply]
The code is separated in more files, most of them should are under GPLV3 since they contain code from the project my work is derivarite from. Two more files in the project directory contain only the MIT licensed code.

About the copyright and license notes. the project im based contained only this:

`# YOLOv5 🚀 by Ultralytics, GPL-3.0 license`

Nothing else, so is that a problem ?


For the MIT licensed code i copied the original copyright and license notes.

**Reply by airtower-luna** [reply]
> About the copyright and license notes. the project im based contained only this:

Well, if that's the original notice that's the original notice. :wink: It doesn't follow the FSF recommendations for what a GPLv3 license notice should look like, but… in my not-a-lawyer opinion it conveys the author's intent. :woman_shrugging: 

Altogether that sounds a lot less troublesome than your previous post, I read that as if you had copied code from both sources into one file. I'm not saying it'd be impossible to do that legally, but more trouble to figure out how to do it right than I'd be willing to go to. :sweat_smile: 

If you have files under MIT license and files under GPLv3, each with their own notices (original and possibly yours for your changes) that's at least pretty similar to how they're doing it for the Linux kernel, and some less prominent software I've seen. Obviously "it's in the kernel" isn't legal advice or law, but… accepted community practice, I suppose. :slightly_smiling_face:

**Reply by zeleny-ondrej** [reply]
thank you so much for your help. The original owner just got back to me saying the notice i've written is quote "perfect" so i guess im good.

**Reply by airtower-luna** [reply]
That's great to hear! And if the copyright owner is happy there's no-one else who could complain. :smile_cat:
